### DISCUSSION TITLE
Can't get raw-window-handle-06 feature to work.

### DISCUSSION THREAD

**Discussion by maxenko** [slint-ui/slint#5319] [Q&A] [answered]
My goal is to get raw window handle so i can check if mouse is in the window rectangle using winapi crate using ```HWND```. Maybe I am just confused about the usage (still new to rust and slint).

my Cargo.toml:
```toml
[dependencies.slint]
version = "1.6.0"
features = ["default", "raw-window-handle-06"]
```

my code (relevant):
```rust
    let ui = MainWindow::new()?;
    let wh = Window::window_handle(&ui.window());
```

This compiles, but I am not seeing ```Result<raw_window_handle_06::WindowHandle<'a>, raw_window_handle_06::HandleError>``` as described in slint api.rs:

```rust
/// This struct represents a persistent handle to a window and implements the
/// [`raw_window_handle_06::HasWindowHandle`] and [`raw_window_handle_06::HasDisplayHandle`]
/// traits for accessing exposing raw window and display handles.
/// Obtain an instance of this by calling [`Window::window_handle()`].
#[cfg(feature = "raw-window-handle-06")]
#[derive(Clone)]
pub struct WindowHandle {
    adapter: alloc::rc::Rc<dyn WindowAdapter>,
}

#[cfg(feature = "raw-window-handle-06")]
impl raw_window_handle_06::HasWindowHandle for WindowHandle {
    fn window_handle<'a>(
        &'a self,
    ) -> Result<raw_window_handle_06::WindowHandle<'a>, raw_window_handle_06::HandleError> {
        self.adapter.window_handle_06()
    }
}

#[cfg(feature = "raw-window-handle-06")]
impl raw_window_handle_06::HasDisplayHandle for WindowHandle {
    fn display_handle<'a>(
        &'a self,
    ) -> Result<raw_window_handle_06::DisplayHandle<'a>, raw_window_handle_06::HandleError> {
        self.adapter.display_handle_06()
    }
}
```
What I am getting is WindowHandle struct.
![window-handle-problem](https://github.com/slint-ui/slint/assets/1999105/9b864cee-3cb3-4838-9dc5-89c43037d701)

How can I get this feature to work and get to raw HWND?

**Comment by tronical** [comment]
You’re almost there :). `let wh = ui.window().window_handle();` gives you the slint::WindowHandle type, which implements the rwh traits. They aren’t implemented on the slint::Window directly because that would be unsafe. The returned WindowHandle type ensures the underlying window stays alive.

**Reply by maxenko** [reply]
@tronical  thank you for the tip, infortunately I am not seeing any special methods, and its still just WindowHandle struct not Result. ie ```let wh = ui.window().```

![window-handle-problem-03](https://github.com/slint-ui/slint/assets/1999105/923c0528-03af-4ea0-b6dc-a8dc4e467aa5)

![window-handle-problem-02](https://github.com/slint-ui/slint/assets/1999105/5d52bdb0-53c2-4ca7-9d6e-3545655ac013)

**Reply by tronical** [reply, marked as answer]
in order to be able to call raw-window-handle’s window_handle function (https://docs.rs/raw-window-handle/latest/raw_window_handle/trait.HasWindowHandle.html#tymethod.window_handle ) on a slint::WindowHandle, you need to bring the HasWindowHandle trait (https://docs.rs/raw-window-handle/latest/raw_window_handle/trait.HasWindowHandle.html# ) in scope. Add raw-window-handle in version 0.6 to your Cargo.toml and add `use raw_window_handle::HasWindowHandle;`

**Comment by maxenko** [comment]
for future readers:
```rust
let ui = MainWindow::new()?;
let slint_wh = ui.window().window_handle();
let maybe_raw_window_handle = slint_wh.window_handle(); // Result<WindowHandle, HandleError>
let mut raw_hwnd = NonZeroIsize::new(-1).unwrap();

match maybe_raw_window_handle {
    Ok(win_handle) => {
        match win_handle.as_ref() {
            raw_window_handle::RawWindowHandle::Win32(handle) => {
                raw_hwnd = handle.hwnd;
            },
            _ => {
                // ...
            }
        }
    },
    Err(e) => {
        // ...
    }
}

```

**Comment by Gibbz** [comment]
I'm having trouble with the import of this into my project.

my toml

```
[dependencies]
slint = "1.9.2"

[build-dependencies]
slint-build = "1.9.2"

[features]
default = [ 
    "slint/default",
    "slint/raw-window-handle-06",
    ]
```

and i get the error on the import:
```
use raw_window_handle::HasWindowHandle;
```
```
error[E0432]: unresolved import `raw_window_handle`
 --> src/main.rs:2:5
  |
2 | use raw_window_handle::HasWindowHandle;
  |     ^^^^^^^^^^^^^^^^^ use of undeclared crate or module `raw_window_handle`

```

**Reply by ogoffart** [reply]
I guess you also need to add the `raw_window_handle` crate as a dependency in your Caro.toml

**Reply by Gibbz** [reply]
thanks, that was it!
